src/utils/config_utils/src/config_utils.c

```diff
@@ -57,13 +58,87 @@ static const char* CONFIG_RUN_AS = "runas";
 static const char* CONFIG_CONNECTION_DATA = "connectionData";
+static const char* CONFIG_CONNECTION_X509_CERT_FILE_PATH = "connectionX509CertFilePath";
+static const char* CONFIG_CONNECTION_X509_PRIVATE_KEY_FILE_PATH = "connectionX509PrivateKeyFilePath";
+static const char* CONFIG_CONNECTION_X509_CA_CERT_FILE_PATH = "connectionX509CaCertFilePath";
 static const char* CONFIG_ADDITIONAL_DEVICE_PROPERTIES = "additionalDeviceProperties";
 static const char* CONFIG_AGENTS = "agents";
 
 static const char* INVALID_OR_MISSING_FIELD_ERROR_FMT = "Invalid json - '%s' missing or incorrect";
 
 static void ADUC_AgentInfo_Free(ADUC_AgentInfo* agent);
 
+
+static char* ADUC_AgentInfo_Read_X509_File(const char* const x509_file_path) {
+    char* buffer = NULL;
+    long ftellResult = 0;
+    size_t bufferLength = 0;
+    size_t bytesRead = 0;
+    FILE* x509File = NULL;
+    
+    x509File = fopen(x509_file_path, "rb");
+    if (x509File)
+    {
+        if (fseek(x509File, 0, SEEK_END) == 0) {
+            ftellResult = ftell(x509File);
+            if (ftellResult != -1)
+            {
+                bufferLength = (size_t)ftellResult;
+                if (fseek(x509File, 0, SEEK_SET) == 0)
+                {
+                    buffer = malloc(bufferLength);
+                    if (buffer)
+                    {
+                        bytesRead = fread(buffer, 1, bufferLength, x509File);
+                        if (bytesRead < bufferLength)
+                        {
+                            if (feof(x509File))
+                            {
+                                Log_Error("Unexpected end of file while reading '%s' (bytesRead: %zu)!", x509_file_path, bytesRead);
+                            }
+                            else if (ferror(x509File)) {
+                                Log_Error("Error occurred while reading '%s' (bytesRead: %zu)!", x509_file_path, bytesRead);
+                            }
+                            else
+                            {
+                                Log_Error("Failed to read '%s' (bytesRead: %zu)!", x509_file_path, bytesRead);
+                            }
+                            free(buffer);
+                            buffer = NULL;
+                        }
+                    }
+                    else
+                    {
+                        Log_Error("Failed to get memory for buffer!");
+                    }
+                }
+                else
+                {
+                    Log_Error("Failed to seek begin of '%s'!", x509_file_path);
+                }
+            }
+            else
+            {
+                Log_Error("Failed to get file position ('%d')!", errno);
+            }
+        }
+        else
+        {
+            Log_Error("Failed to seek end of '%s'!", x509_file_path);
+        }
+
+        if(fclose (x509File) != 0) {
+            Log_Warn("Failed to close '%s'!", x509_file_path);
```

Comment:
omit. no need to log when fclose fails as it's not really actionable and the log trace might not even be able to flush to disk in that case (things are in a pretty bad state)